src/main/java/com/englishtown/bitbucket/hook/MirrorRepositoryHook.java

```diff
@@ -1,262 +1,144 @@
+                                SettingsReflectionHelper settingsReflectionHelper) {
         logger.debug("MirrorRepositoryHook: init started");
 
-        // Set fields
-        this.scmService = scmService;
-        this.i18nService = i18nService;
-        this.executor = executor;
         this.passwordEncryptor = passwordEncryptor;
         this.settingsReflectionHelper = settingsReflectionHelper;
-        this.repositoryService = repositoryService;
 
-        // Init password encryptor
-        PluginSettings pluginSettings = pluginSettingsFactory.createSettingsForKey(PLUGIN_SETTINGS_KEY);
-        passwordEncryptor.init(pluginSettings);
+        int attempts = propertiesService.getPluginProperty(PROP_ATTEMPTS, 5);
+        int threads = propertiesService.getPluginProperty(PROP_THREADS, 3);
+
+        pushExecutor = concurrencyService.getBucketedExecutor(getClass().getSimpleName(),
+                new BucketedExecutorSettings.Builder<>(MirrorRequest::toString, pushProcessor)
+                        .batchSize(Integer.MAX_VALUE) // Coalesce all requests into a single push
+                        .maxAttempts(attempts)
+                        .maxConcurrency(threads, ConcurrencyPolicy.PER_NODE)
+                        .build());
 
         logger.debug("MirrorRepositoryHook: init completed");
     }
 
     /**
-     * Calls the remote bitbucket instance(s) to push the latest changes
-     * <p>
-     * Callback method that is called just after a push is completed (or a pull request accepted).
-     * This hook executes <i>after</i> the processing of a push and will not block the user client.
-     * <p>
-     * Despite being asynchronous, the user who initiated this change is still available from
+     * Schedules pushes to apply the latest changes to any configured mirrors.
      *
-     * @param context    the context which the hook is being run with
-     * @param refChanges the refs that have just been updated
+     * @param context provides any settings which have been configured for the hook
+     * @param request describes the push
      */
     @Override
-    public void postReceive(
-            @Nonnull RepositoryHookContext context,
-            @Nonnull Collection<RefChange> refChanges) {
-
-        logger.debug("MirrorRepositoryHook: postReceive started.");
-
-        List<MirrorSettings> mirrorSettings = getMirrorSettings(context.getSettings());
-
-        for (MirrorSettings settings : mirrorSettings) {
-            runMirrorCommand(settings, context.getRepository());
-        }
-
-    }
-
-    void runMirrorCommand(MirrorSettings settings, final Repository repository) {
-        if (repositoryService.isEmpty(repository)) {
+    public void postUpdate(@Nonnull PostRepositoryHookContext context,
+                           @Nonnull RepositoryPushHookRequest request) {
+        Repository repository = request.getRepository();
+        if (!GitScm.ID.equalsIgnoreCase(repository.getScmId())) {
```

Comment:
This defends against non-Git repositories, rather than doing an `instanceof` check on the `ScmCommandBuilder`. (Not that there are any other supported SCMs)